### PR TITLE
Retrieve Graph Data from DSNP

### PR DESCRIPTION
**Pull request by acruikshank**
Purpose
---------------
In order to know who a user follows (or unfollows), we need to retrieve graph change event from DSNP batch data.

Solution
---------------
This PR retrieves all graph change events from DSNP batches and uses it to populate the graph in Redux. It then uses that to update the UI in various places like the connection list, and the user's feed.

Change summary
---------------
* Alter populate script to add follow information
* Update the SDK to incorporate a parquet issue storing graph change events
* Read graph change events in the SDK
* Alter the redux store to store follows and followers efficiently
* Modify PostLIst to use the new graph change events.
* Modify ConnectionList to use the new graph change events.
* Fix tests.
* Update package-lock.json to version 2 (hopefully for the last time).

Steps to Verify
----------------
1. Start up a clean, deployed chain and the static server.
2. Run the new populate script.
3. Start the app.
4. Confirm that there are posts in the "Feed" tab of the post list.
5. Confirm that you see connections when you click "followers" in the profile block.
6. Confirm that you see connections when you click "following" in the profile block.
7. Confirm that the posts in the feed are consistent with the following list.
8. Confirm that the the follow/unfollow buttons in the "followers" section are consistent with the follow list.

### CODE REVIEW

File: src/components/ConnectionsListProfiles.tsx [hunk 1]
```diff
@@ -13,32 +16,39 @@ enum ListStatus {
...
-    return profiles[userProfile.socialAddress]?.icon?.[0]?.href;
-  };
+  const profileForId = (userId: DSNPUserId): Profile =>
+    profiles[userId] || {
+      ...createProfile({ name: "Annonymous" }),
+      contentHash: "",
+      url: "",
+      announcementType: AnnouncementType.Profile,
+      fromId: userId,
+      handle: "unkown",
```
Comment (enddynayn) [inline src/components/ConnectionsListProfiles.tsx:42]: unkown ===> unknown

**Review by enddynayn — APPROVED** [review]

File: src/services/sdk.ts [hunk 2]
```diff
@@ -244,28 +239,73 @@ const handleRegistryUpdate = (dispatch: Dispatch) => (
...
+    AnnouncementType.GraphChange
+  );
+};
+
+export const isBroadcastAnnouncement = (
+  obj: unknown
+): obj is BroadcastAnnouncement => {
+  const type = (obj as Record<string, unknown>)["announcementType"];
+  return (
+    type === AnnouncementType.Profile ||
```
Comment (enddynayn) [inline src/services/sdk.ts:256]:
```
[ 
  AnnouncementType.Profile,
  AnnouncementType.Broadcast,
  AnnouncementType.Reply
 ].some( (item) => item === type)
```

maybe another way to write this

File: src/services/sdk.ts [hunk 3]
```diff
@@ -244,28 +239,73 @@ const handleRegistryUpdate = (dispatch: Dispatch) => (
...
+
+const fetchAndDispatchContent = (
+  dispatch: Dispatch,
+  message: BroadcastAnnouncement,
+  blockNumber: number
+) => {
+  const decoder = new TextDecoder();
+
+  const url = decoder.decode((message.url as any) as Uint8Array);
+  fetch(url)
```
Comment (enddynayn) [inline src/services/sdk.ts:270]: curios why not use await?

File: scripts/populate.mjs [hunk 4]
```diff
@@ -44,6 +44,7 @@ const accounts = [
         "https://www.dietandi.com/wp-content/uploads/2011/11/Eating-Cereal.jpg"
       ),
     ],
+    follows: [4, 13, 14, 18],
```
Comment (enddynayn) [inline scripts/populate.mjs:47]: Thanks!

**Comment by enddynayn** [comment]
👍 fabolous!

File: scripts/populate.mjs [hunk 5]
```diff
@@ -491,4 +518,42 @@ for await (let account of accounts.values()) {
   await core.contracts.publisher.publish([publication], { signer: wallet });
 }
 
+/**
```
[the remark this answers is not in the document]
Comment (acruikshank) [reply scripts/populate.mjs:521]: I don't think so. It's definitely working. Why do you suggest this?

File: src/components/ConnectionsListProfiles.tsx (hunk 1 above)
Comment (acruikshank) [reply src/components/ConnectionsListProfiles.tsx:42]: Good catch.

File: src/services/sdk.ts (hunk 2 above)
Comment (acruikshank) [reply src/services/sdk.ts:256]: This is nice, but the way it is a a little simpler (a little dumber), so I think I'll keep it.

File: src/services/sdk.ts (hunk 3 above)
Comment (acruikshank) [reply src/services/sdk.ts]: Good question. Fixed.

File: package.json [hunk 6]
```diff
@@ -7,7 +7,7 @@
   "dependencies": {
     "@dsnp/contracts": "1.0.1",
     "@dsnp/parquetjs": "^1.1.0",
-    "@dsnp/sdk": "2.1.0",
+    "@dsnp/sdk": "file:dsnp-sdk-2.1.0.tgz",
```
[the remark this answers is not in the document]
Comment (acruikshank) [reply package.json]: I was missing a commit. It's updated to @dsnp/sdk@next now

File: src/components/ConnectionsList.tsx [hunk 7]
```diff
@@ -17,91 +14,18 @@ const ConnectionsList = (): JSX.Element => {
   const userId: DSNPUserId | undefined = useAppSelector(
     (state) => state.user.id
   );
-  const graphs: Graph[] = useAppSelector((state) => state.graphs.graphs);
-  const graph: Graph | undefined = graphs.find(
-    ({ dsnpUserId }) => dsnpUserId === userId
+  const following = useAppSelector(
+    (state) =>
+      (userId !== undefined ? state.graphs.following[userId] : undefined) || {}
```
[the remark this answers is not in the document]
Comment (acruikshank) [reply src/components/ConnectionsList.tsx]: good call

File: src/components/ConnectionsListProfiles.tsx [hunk 8]
```diff
@@ -13,32 +16,39 @@ enum ListStatus {
...
   const profiles: Record<HexString, Profile> = useAppSelector(
     (state) => state.profiles?.profiles || {}
   );
 
-  const profileIcon = (userProfile: Profile) => {
-    return profiles[userProfile.socialAddress]?.icon?.[0]?.href;
-  };
+  const profileForId = (userId: DSNPUserId): Profile =>
+    profiles[userId] || {
+      ...createProfile({ name: "Annonymous" }),
```
Comment (shannonwells) [inline src/components/ConnectionsListProfiles.tsx:37]:
```suggestion
      ...createProfile({ name: "Anonymous" }),
```

File: src/components/ConnectionsListProfiles.tsx [hunk 9]
```diff
@@ -48,9 +58,9 @@ const ConnectionsListProfiles = ({
           key={userProfile.fromId}
         >
           <UserAvatar
-            icon={profileIcon(userProfile)}
+            icon={profiles[userProfile.fromId]?.icon?.[0]?.href}
```
Comment (shannonwells) [inline src/components/ConnectionsListProfiles.tsx:61]: Are we not going to fallback to blockies?

File: src/components/test/ConnectionsListProfiles.test.tsx [hunk 10]
```diff
@@ -9,55 +9,147 @@ enum ListStatus {
...
+    it("shows all connections user is following", () => {
+      const component = mount(
+        componentWithStore(ConnectionsListProfiles, createMockStore(store), {
+          listStatus: ListStatus.FOLLOWING,
+          followers: followers,
+          following: following,
+        })
+      );
+      expect(component.find(".ConnectionsListProfiles__profile").length).toBe(
+        2
```
Comment (shannonwells) [inline src/components/test/ConnectionsListProfiles.test.tsx:71]: I'd feel better if this compared with following length.  If there's a future bug (e.g. an off-by-1) and somehow this  list got changed this test could possibly pass.  Same with followers.

File: src/test/testGraphs.ts [hunk 11]
```diff
@@ -46,60 +62,31 @@ const adr6 = getPrefabDsnpUserId(6);
...
-      following: [adr0, adr6],
-      followers: [adr0, adr2],
-    },
-    {
-      dsnpUserId: adr2,
-      following: [adr0, adr1, adr3, adr4, adr5, adr6],
-      followers: [],
+export const getPreFabSocialGraph = (): SocialGraph => {
+  const following = {
+    [adr0]: { [adr1]: true, [adr2]: true },
```
Comment (shannonwells) [inline src/test/testGraphs.ts:67]: Can we have one of these, plus a test, with 3 followers? There can be bugs where 1 item or 2 items works but it breaks on  3 items.  3 is more N-like than 1 or 2.

File: src/test/testGraphs.ts [hunk 12]
```diff
@@ -1,18 +1,32 @@
-import { Graph, HexString } from "../utilities/types";
+import { DSNPUserId } from "@dsnp/sdk/dist/types/core/identifiers";
+import { HexString } from "../utilities/types";
 import { generateDsnpUserId, getPrefabDsnpUserId } from "./testAddresses";
 
+type Graph = Record<DSNPUserId, Record<DSNPUserId, boolean>>;
```
Comment (shannonwells) [inline src/test/testGraphs.ts:5]: can you put some more explanatory comments in here? I think a Graph indicates action direction, e.g. in the case of "follower" it means the first DSNPUserId is Following DSNPUserId 2, === true, but I'm not sure.

File: src/test/testGraphs.ts [hunk 13]
```diff
@@ -1,18 +1,32 @@
...
+): Record<DSNPUserId, boolean> => {
+  return {};
+};
 
-  return {
-    dsnpUserId,
-    following,
-    followers,
-  };
+const followersFromFollowing = (following: Graph): Graph => {
```
Comment (shannonwells) [inline src/test/testGraphs.ts:19]: What is the purpose of this function?

**Review by shannonwells — CHANGES_REQUESTED** [review]
Well I had some changes to request but too late

File: src/test/testGraphs.ts (hunk 13 above)
Comment (acruikshank) [reply src/test/testGraphs.ts:19]: Given a map of accounts to a map of the accounts they follow, this inverts the map into a map of followed accounts to a map of accounts that follow them. I'll document this in the next PR.

File: src/test/testGraphs.ts (hunk 12 above)
Comment (acruikshank) [reply src/test/testGraphs.ts:5]: sure.

File: src/test/testGraphs.ts (hunk 11 above)
Comment (acruikshank) [reply src/test/testGraphs.ts:67]: Sure. I'm not sure I agree 3 is significantly more n-like though.

File: src/components/test/ConnectionsListProfiles.test.tsx (hunk 10 above)
Comment (acruikshank) [reply src/components/test/ConnectionsListProfiles.test.tsx:71]: Ok. next PR.

File: src/components/ConnectionsListProfiles.tsx (hunk 9 above)
Comment (acruikshank) [reply src/components/ConnectionsListProfiles.tsx:61]: We fall back to blockies in the UserAvatar if this value is undefined.

File: src/components/ConnectionsListProfiles.tsx (hunk 8 above)
Comment (acruikshank) [reply src/components/ConnectionsListProfiles.tsx:37]: Good catch. Next PR.

**Pull request merged**